### ISSUE TITLE
useless_asref false positive on `&mut &[T]`

### ISSUE THREAD

**Issue by notpeelz** [rust-lang/rust-clippy#8566]
### Summary

Clippy complains about calling `as_ref()` on a mut slice when it shouldn't.

### Lint Name

useless_asref

### Reproducer

```rs
use std::{error::Error, io::BufRead};

fn main() {
  let buf = get_buf();

  // clippy complains with: "this call to `as_ref` does nothing"
  test(&mut buf.as_ref()).unwrap();

  // this is fine
  test(&mut &*buf).unwrap();
}

fn get_buf<'a>() -> &'a mut [u8] {
  todo!()
}

fn test<R>(_input: &mut R) -> Result<(), Box<dyn Error>>
where
  R: BufRead + Send + Sync,
{
  Ok(())
}
```

### Version

```text
rustc 1.59.0 (9d1b2106e 2022-02-23)
binary: rustc
commit-hash: 9d1b2106e23b1abd32fce1f17267604a5102f57a
commit-date: 2022-02-23
host: x86_64-unknown-linux-gnu
release: 1.59.0
LLVM version: 13.0.0
```

**Comment by giraffate** [comment]
It's related to https://github.com/rust-lang/rust-clippy/issues/5862.
